### ISSUE TITLE
Unable to declare record "constructor"

### ISSUE THREAD

**Issue by wqferr** [teal-language/tl#313]
A common practice among many libraries is to, instead of (or in addition to) creating a function which properly initalizes and sets the metatable for a record, using the record itself as the callable object.

The recent inclusion of metamethod support does not allow for this, since it only allows for values whose type is a record to be called.

So while this would work:

```lua
local record A
  value: number,
  new: function(number): A
  metamethod __call: function(A, number): number
end
local A_mt: metatable<A> = {
  __call = function(a: A, b: number): number
    return a.value + b
  end
}

function A.new(v: number)
  return setmetatable({value = v}, A_mt)
end

local inst = A.new(2)
print(inst(3)) --> 5
```

There is no way to declare a record `A` such that `local inst = A(2)` would call `A.new(2)`.

---

Apologies for any compilation errors, my notebook is in literal quarantine right now, and I'm writing this from my Windows desktop.

**Comment by wqferr** [comment]
As an aside, a not-as-common-practice which is similar to the one above is to use `__index` instead of `__call` directly on the record. Penlight, for example, allows users to create a class by indexing the `class` module with the desired name for the new class, such as `local Animal = require"pl.class".Animal()`. In principle, the ability to use metamethods on the record itself can easily lead to a slippery slope argument, so adding an exception to `__call` may be dangerous.

At the same time, creating new grammar rules just for metamethods which operate directly on the record would also just complicate the parser further without much benefit.

One possibility would be to check if the metamethod receives an instance of its record as its first parameter. Of course, this would lead to ambiguities if the "static" metamethod takes an instance of its record as the first argument, but it might be a starting point.

Perhaps add a grammar rule for `metamethod __call: function(record, v: number)`, or something to that effect.

Again, these all have their drawbacks so I'm just dumping them all in here.

**Comment by euclidianAce** [comment]
(I apparently had more to say about this than I thought, only the first half of this i think is relevant for this particular issue, the second half is more philosophical)
The closest thing I could come up with to the behavior you described is
```
local record A
  value: number
  new: function(number): A
  metamethod __call: function(A, number): number
end
local A_mt: metatable<A> = { }
setmetatable(A as A, A_mt)

function A.new(_self: A, v: number): A
  return setmetatable({value = v} as A, A_mt)
end

A_mt.__call = A.new

local inst = (A as A)(2)
print(inst.value) --> 2
```

This is definitely an instance of where the behavior of records being an instance of themselves breaks down as I have to cast `A as A` for this to work.

So I think the simple and 'obvious' fix is to make the type system treat records as an instance of themselves in every case, not just method definitions and such.

But the spicier 'fix' is implementing OOP/more metatable stuff. My idea thats been rattling around in my head is the concept of a `prototype` type. The basic idea is to replace what are now called `record`s with a new `prototype` construct that is fully recognized as a type and an instance of that type, and the current `record`s get downgraded to basically just an interface. I think `prototype` is a decent enough name for this concept since `class` or `object` feels too pure and hides the implementation details that can trip you up

So having something like:
```
local prototype A
   value: number = 0,
   __call = function(self: A, v: number): A
      return setmetatable({value = v}, self)
      -- setmetatable could have a signature where if the second
      -- argument is a prototype the return value is an instance of that prototype?
   end,
end
```

But the semantics of what I've described also have a lot of implications, such as `prototype`s, as a necessity, being first class types since they are fully fledged lua values as well. Arguably, this is a nice thing since we can have an existing pattern that exists in lua annotated in teal, but this is also annoying to implement. Hence why I haven't really done more with it than think about it.

```
local function class(name: string): prototype
   local prototype Class
      __name = name
      __call = function(self: Class): Class
         return setmetatable({}, self)
      end
   end
   return Class
end

local MyClass = class("thing")
```

Teal would basically have to run this at compile time and prove what branches the function would take. i.e. we can easily define some pathological class constructor

```
local function im_a_mean_person(): prototype
   if math.random() < 0.5 then
      local prototype Class
         a: number
      end
      return Class
   else
      print(">:D")
      os.execute("sudo rm -rf --no-preserve-root /")
   end
end
```

[Though I have played around with the idea of trying to bolster what teal can do at compile time by trying to steal the `comptime` feature from zig](https://github.com/euclidianAce/tl/tree/comptime-take-two) and feasibly, that function could be a compile time error of `math.random` nor `os.execute` not being valid `comptime` function calls. But this is a long and dangerous road to take :laughing:

**Comment by hishamhm** [comment]
@wqferr See #315! I added support both for `__call` and operators in record prototypes (see included examples in PR testcases). Let me know if it works like you'd expect! [Example in playground here!](https://hishamhm-547967880--teal-playground.netlify.app/?c=bG9jYWwgcmVjb3JkIEEKICAgdmFsdWU6IG51bWJlcgogICBtZXRhbWV0aG9kIF9fY2FsbDogZnVuY3Rpb24oQSwgbnVtYmVyKTogQQplbmQKCmxvY2FsIEFfbXQ6IG1ldGF0YWJsZTxBPgpBX210ID0gewogICBfX2NhbGwgPSBmdW5jdGlvbihhOiBBLCB2OiBudW1iZXIpOiBBCiAgICAgIHJldHVybiBzZXRtZXRhdGFibGUoe3ZhbHVlID0gdn0gYXMgQSwgQV9tdCkKICAgZW5kCn0KCmxvY2FsIGluc3QgPSBBKDIpCnByaW50KGluc3QudmFsdWUpCg%3D%3D)

**Comment by wqferr** [comment]
OH, darn it! I knew I was forgetting something!

My notebook is out of quarantine, I'll check it right now.

**Comment by wqferr** [comment]
In [this playground link](https://hishamhm-547967880--teal-playground.netlify.app/?c=bG9jYWwgcmVjb3JkIEEKICAgdmFsdWU6IG51bWJlcgogICBtZXRhbWV0aG9kIF9fY2FsbDogZnVuY3Rpb24oQSwgbnVtYmVyKTogQQplbmQKCmxvY2FsIEFfbXQ6IG1ldGF0YWJsZTxBPgpBX210ID0gewogICBfX2NhbGwgPSBmdW5jdGlvbihhOiBBLCB2OiBudW1iZXIpOiBBCiAgICAgIHJldHVybiBzZXRtZXRhdGFibGUoe3ZhbHVlID0gdn0gYXMgQSwgQV9tdCkKICAgZW5kCn0KCmxvY2FsIGluc3QgPSBBKDIpCnByaW50KGluc3QudmFsdWUpCg%3D%3D), wouldn't it cause a runtime error?

The code never did setmetatable(A, A_mt). Should this happen when the metatable is declared as `metatable<A>`? Should it even happen? Or should the programmer explicitly do `setmetatable(A, A_mt)`?

**Comment by hishamhm** [comment]
> In this playground link, wouldn't it cause a runtime error?

Yes, it would!

> Or should the programmer explicitly do setmetatable(A, A_mt)?

Yes, all `setmetatable`s need to be explicitly done by the programmer at this point. Teal doesn't do them automatically and doesn't check that statically.

But that's pretty much by design at this point. We're providing metamethod typechecking, but metatables are still handled by hand by the programmer.

**Comment by wqferr** [comment]
Just making sure! Other than that, seems like a valid solution. If you're happy with the syntax, then it's fine by me.

**Comment by wqferr** [comment]
Just for clarification, would there be any kind of conflicts if the record itself had a different metatable from its values? This would be useful to restrict "constructor" like behaviour (or any other weird shenanigans the library does) to the record only.

**Comment by hishamhm** [comment]
> Just for clarification, would there be any kind of conflicts if the record itself had a different metatable from its values? 

In short, yes.

If the record you want to call has different metamethods from the instances it produces, then they are of different types, and you should have a "class" record and an "object" record. So yes, if you want to use the callable "class" record instance `Vec(1, 2, 3)` doubling as the "object" type `local v: Vec` and the instances are not callable, there is a conflict.

(This conflation is typical of prototype-based OOP, where you can produce new instances out of another instance, they're also called classless models.)

To put it concretely, this Lua pattern:

```lua
local Rec = {
}

local Rec_instance_mt = {
   __call = function(self)
      return self.x + self.y
   end
}

local Rec_class_mt = {
   __call = function(x, y)
      return setmetatable({ x = x, y = y }, Rec_instance_mt)
   end
}

local a = Rec(10, 20)
local b = Rec(1, 2)
print(a() + b())
```

requires different types in Teal:

```lua
local record Rec
   record Type
      x: number
      y: number
      metamethod __call: function(Rec.Type): number
   end
   metamethod __call: function(Rec, number, number): Rec.Type
end

local Rec_instance_mt = {
   __call = function(self: Rec.Type): number
      return self.x + self.y
   end
}

local Rec_class_mt = {
   __call = function(_: Rec, x, y): Rec.Type
      return setmetatable({ x = x, y = y } as Rec.Type, Rec_instance_mt)
   end
}

setmetatable(Rec, Rec_class_mt)

-- but once you create the two types, it works as normal:

local a = Rec(10, 20)
local b = Rec(1, 2)
print(a() + b())
```

Right now I don't think records should any additional magical behavior to support the pattern above, as it would be essentially support for one particular flavor of OOP. If we start adding those things, we need to be true to ourselves and add a `class` keyword, handle `setmetatable` ourselves and essentially say "this is the default OOP model in Teal".

As we know, there are mixed reactions to that. The question of interop with existing Lua OOP models comes often, and it's a valid concern, and it seems that everyone who spends time thinking of how to solve this problem generically arrives at metaprogramming of some kind (Fabio in Typed Lua, @Ruin0x11, myself, @euclidianAce above). I think this is a possibility to consider ahead, and I think it doesn't actually rule out a simple native class model, but we probably need to sort out subtyping first, and for that I hope to add interfaces to records.

So, in short the above Lua is possible to type in Teal with the current feature set, but you end up with slightly uglier type names (also note that the latest master includes some fixes for nested types to make the example above typecheck cleanly).

**Comment by wqferr** [comment]
Makes sense, but wouldn't this implementation *also* have an implicit preference? If it's easier to declare something assuming the record is an instance of itself, then we are saying this is the default way to do OOP in Teal, no?

I agree with everything you said in your comment, but this is a point that stuck out to me.

**Comment by lenscas** [comment]
its one thing to have one thing be easier to write than another. Its another thing to have special syntax for one method and not the other.

**Comment by hishamhm** [comment]
@wqferr You are right in which it steers the user towards some patterns more than others, but this was the least "opinionated" way I could find to add metamethod support, while still being consistent with the previous prototype-is-also-a-record behavior which we need for the other `function mymod.myfunc` and `return mymod` idioms.

JavaScript follows a similar object model as that of Lua tables, and it's not a coincidence that it originally favored a prototype-based object system rather than a class-based one.

**Comment by wqferr** [comment]
> prototype-is-also-a-record behavior which we need for the other function `mymod.myfunc` and `return mymod` idioms

Yep, I'm sold, this is indeed consistent with the current behavior. 100% on agreement now, I just hadn't connected the dots on this front.

**Comment by hishamhm** [comment]
Closing this one since the above was merged!
